### PR TITLE
perf(ivy): chain listener instructions

### PR DESCRIPTION
**Pull request by crisbeto**
Chains multiple listener instructions on a particular element into a single call which results in less generated code. Also handles listeners on templates, host listeners and synthetic host listeners.

### CODE REVIEW

**Review by pkozlowski-opensource — APPROVED** [review]
Why not, LGTM for runtime.

> results in less generated code

Did you happen to have example numbers from the apps you are working on?

File: packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts [hunk 1]
```diff
@@ -1134,6 +1134,114 @@ describe('compiler compliance: bindings', () => {
...
+            if (rf & 1) {
+              $r3$.ɵɵcomponentHostSyntheticListener("@animation.done", function MyComponent_animation_animation_done_HostBindingHandler($event) { return ctx.done(); })("@animation.start", function MyComponent_animation_animation_start_HostBindingHandler($event) { return ctx.start(); });
+            }
+          }
+        `;
+
+      expectEmit(result.source, template, 'Incorrect template');
+    });
+
+    it('should chain multiple regular and synthetic host listeners into a single instruction',
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts]: Is this test description accurate? It looks to me like we get two instructions.

**Review by petebacondarwin — APPROVED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts (hunk 1 above)
Comment (crisbeto) [reply packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts]: Good point, that's what I was trying to say.

**Comment by crisbeto** [comment]
@pkozlowski-opensource I don't have numbers for these changes because I don't have access to my app this week, but the approach is similar to the other chained instructions which showed improvements in bundle size.

File: packages/core/src/render3/instructions/listener.ts [hunk 2]
```diff
@@ -68,11 +69,12 @@ export function ɵɵlistener(
 */
 export function ɵɵcomponentHostSyntheticListener(
     eventName: string, listenerFn: (e?: any) => any, useCapture = false,
-    eventTargetResolver?: GlobalTargetResolver): void {
+    eventTargetResolver?: GlobalTargetResolver): TsickleIssue1009 {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/listener.ts]: Quick comment: it looks like the [mentioned Tsickle issue is resolved](https://github.com/angular/tsickle/issues/1009) :)
We can probably create a followup "cleanup" PR to use real types here and in other places.

**Comment by AndrewKushnir** [comment]
- [VE Presubmit](http://test/OCL:279773962:BASE:279774032:1573499056168:efa58a0f)
- [Ivy Presubmit](http://test/OCL:279773962:BASE:279774291:1573499116069:a7e79f1a)

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/instructions/listener.ts (hunk 2 above)
Comment (crisbeto) [reply packages/core/src/render3/instructions/listener.ts]: Ah nice, I can do it once I'm done adding chaining to some of the other instructions.

**Pull request closed without merging**